### PR TITLE
Check argument types in contract calls (RFC)

### PR DESCRIPTION
**Pull request by gorillainduction**
Just a first progress PR. I don't intend to merge this, but I welcome comments on the approach.

### CODE REVIEW

**Comment by hanssv** [comment]
I think the approach looks sound :-)

File: apps/aecontract/src/aect_dispatch.erl [hunk 1]
```diff
@@ -179,3 +191,45 @@ chain_state(#{ contract    := ContractPubKey
...
+                        {ok, GivenArgType} -> ok;
+                        {ok, ArgType} ->
+                            %% Check if both type reps give the same result
+                            ?DEBUG_LOG("Checking representation for ~p\n"
+                                       "Given   : ~p\n"
+                                       "Accepted: ~p"
+                                      , [Function, GivenArgType, ArgType]),
+                            ExpectedType = {tuple, [string, ArgType]},
+                            case aeso_data:from_binary({tuple, [typerep, ExpectedType]}, CallData) of
+                                {ok, {_, {_, Args}}} -> ok;
```
Comment (UlfNorell) [inline apps/aecontract/src/aect_dispatch.erl]: Is this what we want? In particular it allows one to specify incorrect types of the constructors of a datatype that you aren't using. This is harmless, I suppose, but it's not clear why we'd want to allow it.

File: apps/aesophia/src/aeso_compiler.erl [hunk 2]
```diff
@@ -84,12 +93,34 @@ to_bytecode([Op|Rest], Options) ->
     [aeb_opcodes:m_to_op(Op)|to_bytecode(Rest, Options)];
 to_bytecode([], _) -> [].
 
+extract_type_info(#{functions := Functions} =_Icode) ->
+    io:format("Functions: ~p\n", [Functions]),
+    TypeInfo = [{list_to_binary(Name),
+                 aeso_data:to_binary({tuple, [T || {_, T} <- Args]}),
+                 aeso_data:to_binary(TypeRep)}
+                || {Name,_Attrs, Args,_Body, TypeRep} <- Functions,
```
Comment (UlfNorell) [inline apps/aesophia/src/aeso_compiler.erl]: Should only extract type info for public functions:
```suggestion
                || {Name, Attrs, Args,_Body, TypeRep} <- Functions,
                   not lists:member(private, Attrs),
```

**Comment by gorillainduction** [comment]
I'm dropping this in favor of the rebased PR https://github.com/aeternity/epoch/pull/1696

**Pull request closed without merging**